### PR TITLE
[SofaEditor] ADD the SofaEditor module to exchange the state of sofa editor.  

### PR DESCRIPTION
**Pull request by damienmarchal**
In this PR I'm adding a module implementing an API which allows the different editors (runSofa, runSofa2, other tool) to communicate their states (eg: current selection) with controller (pythonscriptcontroller) & co. 

The API is the following:
```cpp
    /// This should be in the gui
    shared_ptr<SofaEditorState> state = make_shared<shared_ptr>({"main"}) ;
    SofaEditor::ID id = SofaEditor::createId(state) ; 
    state->setSelectionFromPath({"/root/child1/object", "/root/child1/object2"})
   
    //// This should be on the client side
   SofaEditor::ID id = SofaEditor::getIdFromEditorName("main");
   shared_ptr<SofaEditorState> s = SofaEditor::getState(id);
   auto& v = s->getSelection(); 
```

It is more common to have the client side in a PythonScriptController. To make that possible this module have a python binding which allows to write interesting editing interaction in python:
```python
from SofaEditor import SofaEditor

def onDraw():
      s = SofaEditor.getSelection()
      drawSomethingSpecialForSelectedObject(s)
```

The python binding is different to the one we have yet into Sofa s it is made totally independently to  SofaPython and is done using cython (http://cython.org/) 
(for which I'm curious to see how this will survive our CI). 

Other notable things are that there is unittest either on the c++ side & python side and that the CMakeLists.txt are installing everything into a single the install location with the following structure: 
module/SofaEditor/SofaEditor.so
module/SofaEditor/python/SofaEditor/__init__.py
module/SofaEditor/python/SofaEditor/SofaEditor.so 
module/SofaEditor/examples
(I'm not sure it is a good idea but this is a kind of "package" but I think it is an interesting way to ship plugins). 

I will open other PR to integrate this feature into runSofa and runSofa2. 

For the records, I was hesitating between three designs. The first one consists in using the "context/ExecParams" to hold the editor state. As this context is passed to the components during the scenegraph traversal this would allows them to know the status of the editor. The second approach was using flexible messages/event transmitted through the scene graph. The third design was much simpler and simply consist of using a singleton so that it can be used everywhere there is a need to know the editor status. Between the three I made the decision to use the simplest approach (simpler to implement & simpler to use) despite he is more limited. 

______________________________________________________

This PR: 
- [ ] builds with SUCCESS for all platforms on the CI.
- [ ] does not generate new warnings.
- [ ] does not generate new unit test failures.
- [ ] does not generate new scene test failures.
- [ ] does not break API compatibility.
- [ ] is more than 1 week old (or has fast-merge label).

**Reviewers will merge only if all these checks are true.**

### CODE REVIEW

**Comment by guparan** [comment]
This PR depends on https://github.com/sofa-framework/sofa/pull/734 and https://github.com/sofa-framework/sofa/pull/732

**Comment by epernod** [comment]
Sorry, I'm not sure to fully understand.
So for example if you are doing some mesh picking using the mouse, the state will returns "picked"?

What is the list of state?

Why the use of Id from getIdFromEditorName ?

**Review by epernod — COMMENTED** [review]
Sorry, I'm not sure to fully understand.
So for example if you are doing some mesh picking using the mouse, the state will returns "picked"?

What is the list of state?

Why the use of Id from getIdFromEditorName ?

**Comment by damienmarchal** [comment]
Hello Erik, 

Thanks for you interest on this PR. 

The general design is to allow 'n' editor, each of them being associated to 1 editorstate. The states can be overiden to add more properties in addition to the selection set and the editor name. 

That's said, for my use case I'm only interested with having one editor connected to one state. 
In this PR I don't change runSofa/runSofa2 so that they fill the editor's state. But yes the basic actions on the editors are to fill the selection when you select a group of nodes/object in the scene graph and when you select by picking in the 3d view. 

EDIT: The ID is there just because I want to be able to connect an editor to a meaningfull name ("main", "qteditor", "insceneeditor") and be able to retrieve them without having to pass it at each call to the getState function. 

EDIT2: The state is currently holding the "selected objects" and so it is not storing "picked" but that "/root/childnode1/child2/objectA" is the current selection. To hold per object property I think the object Data are good enough for that.

**Comment by epernod** [comment]
ok understood for the Id. It makes totally sens. 

What is the SofaEditor::  ? it would be for example runSofa, runSofa2 ?
And I can't find the list of state for your python editor in the Pr changes.

File: modules/SofaEditor/python/tests/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+project(PythonSofaEditor_test)
+
+####################################################################################################
+### Module dependencies
+####################################################################################################
+find_package(SofaPython 2.7 QUIET)
+if(!SofaPython_FOUND)
+    message(FATAL_ERROR "-- 'PythonSofaEditor_test' is disabled because 'SofaPython' is missing or not activated.")
+    return()
```
Comment (guparan) [inline modules/SofaEditor/python/tests/CMakeLists.txt]: equivalent to `find_package(SofaPython 2.7 REQUIRED)` ?

Comment (damienmarchal) [reply modules/SofaEditor/python/tests/CMakeLists.txt]: I don't think so. 
REQUIRED will make the cmake process to fail while the pattern I'm using is just not doing the python binding.

Comment (guparan) [reply modules/SofaEditor/python/tests/CMakeLists.txt]: `message(FATAL_ERROR` will make the CMake process fail

**Comment by damienmarchal** [comment]
@guparan What about this PR ?

**Comment by damienmarchal** [comment]
[ci-build][with-scene-tests]

**Comment by damienmarchal** [comment]
@guparan, @fredroy, @marques-bruno 

I'm facing a problem here, all the scene are failing because of interaction between the library in this PR and the "automatically loading plugin" mechanism we have. 

In this PR, a new module is added but as this does not contains sofa components there is no need for to implement the 'plugin' interface (the initExternalModule() and other functions). But, by default the build scripts are automatically adding all plugins&modules into the config file with the plugins to load by automatically including things that does not have components (and thus are "simple shared libraries). 

What do you think about:
- differentiate better the 'libraries' from the plugins. This may be by adding adding a 
  sofa_add_library() ? 
- store the libraries in a different build location than the plugins/modules. 
- I can add the initSofaEditor.cpp with the missing function...it would work but as this does make sens I'm hesitating to such solution.

**Comment by marques-bruno** [comment]
@damienmarchal I think it really depends on how many of those libraries (code without components to register to the factory, or external modules to load) we expect to see in the future.
Since it's a one-time loading thing, It doesn't seem very costly to load and call the plugin interface in this case. but if we want to democratize this kind of modules, it might be worth modifying the API.
Also, even for a library, a getModuleDescription / getModuleName / license has some value

**Comment by damienmarchal** [comment]
The PR is deprecated since plugin.SofaPython3.

**Pull request closed without merging**